nuxeo-compound-documents/src/main/java/org/nuxeo/compound/documents/service/CompoundDocumentServiceImpl.java

```diff
@@ -16,21 +16,93 @@
+
+    private static final Set<String> BLACKLIST = Set.of("__MACOSX", "../", ".DS_Store", "desktop.ini", "Thumbs.db");
+
     @Override
-    public DocumentModel createCompoundDocument(DocumentModel parent, Blob archiveBlob) {
-        String compoundDocName = FilenameUtils.removeExtension(archiveBlob.getFilename());
+    public DocumentModel createCompoundDocument(DocumentModel parent, Blob archive) throws IOException {
+        String compoundDocName = FilenameUtils.removeExtension(archive.getFilename());
         String parentPath = parent.getPathAsString();
         CoreSession session = parent.getCoreSession();
-        return session.createDocument(session.createDocumentModel(parentPath, compoundDocName, "CompoundDocument"));
+        DocumentModel compoundDoc = session.createDocument(
+                session.createDocumentModel(parentPath, compoundDocName, "CompoundDocument"));
+        createFolderStructure(compoundDoc, archive);
+        populateFolderStructure(compoundDoc, archive);
+        return compoundDoc;
+    }
+
+    protected void createFolderStructure(DocumentModel compoundDoc, Blob archive) throws IOException {
+        try (ZipFile zip = new ZipFile(archive.getFile())) {
+            zip.stream()
+               .filter(this::isAllowedEntry)
+               .filter(ZipEntry::isDirectory)
+               .sorted(Comparator.comparing(ZipEntry::getName))
+               .forEach(e -> createCompoundDocumentFolder(compoundDoc, e.getName()));
+        }
+    }
+
+    protected boolean isAllowedEntry(ZipEntry entry) {
+        Path path = Paths.get(entry.getName());
+        Iterator<Path> iterator = path.iterator();
+        while (iterator.hasNext()) {
+            Path fragment = iterator.next();
+            if (BLACKLIST.contains(fragment.toString()) || fragment.startsWith(".")) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    protected DocumentModel createCompoundDocumentFolder(DocumentModel compoundDoc, String name) {
+        CoreSession session = compoundDoc.getCoreSession();
+        Path path = Paths.get(compoundDoc.getPathAsString(), name);
+        return session.createDocument(session.createDocumentModel( //
+                path.getParent().toString(), path.getFileName().toString(), "CompoundDocumentFolder"));
+    }
+
+    protected void populateFolderStructure(DocumentModel compoundDoc, Blob archive) throws IOException {
+        FileManagerService fileManager = (FileManagerService) Framework.getService(FileManager.class);
+        try (ZipFile zip = new ZipFile(archive.getFile())) {
+            zip.stream().filter(this::isAllowedEntry).filter(Predicate.not(ZipEntry::isDirectory)).forEach(entry -> {
+                Path path = Paths.get(compoundDoc.getPathAsString(), entry.getName());
+                try {
+                    FileBlob blob = new FileBlob(zip.getInputStream(entry));
```

Comment:
Replace by a FileBlob by a ZipEntryBlob (which will close the stream automatically and not copy the zip on disk (FileBlob makes a temp file))